File: packages/agent/agent/tools/list_documents.ts
import { getDb } from "@clara-financas/db";
import { extractionStagings } from "@clara-financas/db/schema/extraction-staging";
import { batches, documents } from "@clara-financas/db/schema/ledger";
import { forTenant } from "@clara-financas/db/tenant-scope";
import { desc, eq, inArray } from "drizzle-orm";
import { defineTool } from "eve/tools";
import { z } from "zod";

import { requireTenantCaller } from "../lib/tenant";

/**
 * Todos os documentos — inclusive os que nenhuma outra lista mostra.
 *
 * `list_invoices` e o snapshot fazem `innerJoin` com `batches`: um documento
 * cuja extração falhou, ou cujo lote foi rejeitado, simplesmente não existia
 * para a conversa. E como o hash barra o reenvio do mesmo PDF, o documento
 * órfão ficava num limbo inalcançável — a pessoa perguntava "cadê a fatura que
 * eu mandei?" e a resposta honesta era inexprimível.
 *
 * Aqui o join é aberto: cada documento aparece com o estado do seu lote mais
 * recente (ou nenhum), e com a extração estacionada quando existe — que é o
 * que permite retomar o fluxo de onde ele parou.
 */
const LIMIT = 50;

export default defineTool({
  description:
    "Lists uploaded documents INCLUDING the ones no invoice list shows: extraction failed (no batch) or batch rejected. Each row carries the latest batch state and any staged extraction. Use for 'cadê o documento que enviei' and to resume a stalled upload — re-uploading the same PDF is blocked by hash, this is how you find it again.",
  inputSchema: z.object({
    withoutBatch: z
      .boolean()
      .optional()
      .describe("true = only documents that have no batch at all (extraction never proposed)."),
    limit: z.number().int().min(1).max(LIMIT).optional(),
  }),
  async execute(input, ctx) {
    const { tenantId } = requireTenantCaller(ctx);
    const limit = input.limit ?? 20;

    return forTenant(
      tenantId,
      async (tx) => {
        const rows = await tx
          .select({
            id: documents.id,
            issuer: documents.issuer,
            kind: documents.kind,
            uploadedAt: documents.uploadedAt,
          })
          .from(documents)
          .where(eq(documents.tenantId, tenantId))
          .orderBy(desc(documents.uploadedAt))
          .limit(LIMIT + 1);

        const truncatedAll = rows.length > LIMIT;
        const page = truncatedAll ? rows.slice(0, LIMIT) : rows;
        const ids = page.map((row) => row.id);

        const [batchRows, stagedRows] =
          ids.length === 0
            ? [[], []]
            : await Promise.all([
                tx
                  .select({
                    id: batches.id,
                    documentId: batches.documentId,
                    status: batches.status,
                    checksumResult: batches.checksumResult,
                    createdAt: batches.createdAt,
                  })
                  .from(batches)
                  .where(inArray(batches.documentId, ids))
                  .orderBy(desc(batches.createdAt)),
                tx
                  .select({
                    id: extractionStagings.id,
                    documentId: extractionStagings.documentId,
                  })
                  .from(extractionStagings)
                  .where(inArray(extractionStagings.documentId, ids)),
              ]);

        // O lote mais recente conta a situação atual; um `rejected` antigo não
        // esconde um `proposed` novo.
        const latestBatch = new Map<string, (typeof batchRows)[number]>();
        for (const batch of batchRows) {
          if (!latestBatch.has(batch.documentId)) latestBatch.set(batch.documentId, batch);
        }
        const stagedByDocument = new Map(stagedRows.map((row) => [row.documentId, row.id]));

        const described = page.map((row) => {
          const batch = latestBatch.get(row.id);
          const stagedExtractionId = stagedByDocument.get(row.id);
          return {
            documentId: row.id,
            displayLabel: row.issuer ?? "Documento enviado",
            issuer: row.issuer,
            kind: row.kind,
            uploadedAt: row.uploadedAt.toISOString(),
            batch:
              batch === undefined
                ? null
                : { batchId: batch.id, status: batch.status, checksumResult: batch.checksumResult },
            ...(stagedExtractionId !== undefined ? { stagedExtractionId } : {}),
            // O que destrava cada estado, dito na linha — para o modelo não
            // ter de deduzir o próximo passo de um join.
            next:
              batch === undefined
                ? stagedExtractionId !== undefined
                  ? "extração pronta e sem lote: propose_batch_from_extraction"
                  : "sem lote e sem extração: delegue ao extrator"
                : batch.status === "rejected"
                  ? "lote descartado: para reaproveitar, delegue ao extrator e proponha de novo"
                  : null,
          };
        });

        const filtered =
          input.withoutBatch === true
            ? described.filter((row) => row.batch === null)
            : described;

        return {
          count: filtered.length,
          ...(truncatedAll
            ? {
                truncated: true as const,
                note: `Mostrando os ${LIMIT} documentos mais recentes; há mais.`,
              }
            : {}),
          documents: filtered.slice(0, limit),
        };
      },
      getDb(),
    );
  },
});

File: services/agent/clara/tools/list_documents.py
"""Todos os documentos — inclusive os que nenhuma outra lista mostra.

Porta de `agent/tools/list_documents.ts`. `list_invoices` e o snapshot fazem
inner join com `batches`: um documento cuja extração falhou, ou cujo lote foi
rejeitado, não existia para a conversa. Aqui o join é aberto.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from sqlalchemy import select
from sqlalchemy.orm import Session

from clara.db.models import Batch, Document, ExtractionStaging

_HARD_LIMIT = 50


@dataclass
class DocumentRow:
    document_id: str
    display_label: str
    issuer: str | None
    kind: str
    uploaded_at: str
    batch: dict[str, Any] | None
    next: str | None
    staged_extraction_id: str | None = None


@dataclass
class ListDocumentsResult:
    count: int
    documents: list[DocumentRow] = field(default_factory=list)
    truncated: bool = False
    note: str | None = None


def list_documents(
    session: Session, tenant_id: str, *, without_batch: bool = False, limit: int = 20
) -> ListDocumentsResult:
    rows = session.execute(
        select(Document)
        .where(Document.tenant_id == tenant_id)
        .order_by(Document.uploaded_at.desc())
        .limit(_HARD_LIMIT + 1)
    ).scalars().all()

    truncated_all = len(rows) > _HARD_LIMIT
    page = rows[:_HARD_LIMIT] if truncated_all else rows
    ids = [r.id for r in page]

    latest_batch: dict[str, Batch] = {}
    staged_by_document: dict[str, str] = {}
    if ids:
        batch_rows = session.execute(
            select(Batch).where(Batch.document_id.in_(ids)).order_by(Batch.created_at.desc())
        ).scalars().all()
        # O lote mais recente conta a situação atual; um `rejected` antigo não
        # esconde um `proposed` novo.
        for batch in batch_rows:
            latest_batch.setdefault(batch.document_id, batch)

        staged_rows = session.execute(
            select(ExtractionStaging.id, ExtractionStaging.document_id).where(
                ExtractionStaging.document_id.in_(ids)
            )
        ).all()
        staged_by_document = {doc_id: sid for sid, doc_id in staged_rows}

    described: list[DocumentRow] = []
    for row in page:
        batch = latest_batch.get(row.id)
        staged_id = staged_by_document.get(row.id)

        if batch is None:
            next_step = (
                "extração pronta e sem lote: propose_batch_from_extraction"
                if staged_id is not None
                else "sem lote e sem extração: delegue ao extrator"
            )
        elif batch.status == "rejected":
            next_step = "lote descartado: para reaproveitar, delegue ao extrator e proponha de novo"
        else:
            next_step = None

        described.append(
            DocumentRow(
                document_id=row.id, display_label=row.issuer or "Documento enviado",
                issuer=row.issuer, kind=row.kind, uploaded_at=row.uploaded_at.isoformat(),
                batch=(
                    None if batch is None
                    else {
                        "batch_id": batch.id,
                        "status": batch.status,
                        "checksum_result": batch.checksum_result,
                    }
                ),
                staged_extraction_id=staged_id, next=next_step,
            )
        )

    filtered = [d for d in described if d.batch is None] if without_batch else described

    return ListDocumentsResult(
        count=len(filtered),
        documents=filtered[:limit],
        truncated=truncated_all,
        note=(
            f"Mostrando os {_HARD_LIMIT} documentos mais recentes; há mais."
            if truncated_all else None
        ),
    )
